### PR TITLE
[Al Amin Step - 13] Search function with Status and Title

### PR DESCRIPTION
**Pull request by rr-mdalauddin-alamin**
### Step 13: Add status to make it searchable

- Let's add status (not started / started / completed)
  - [Optional requirements] If you are not a beginner, you may install a Gem that manages the state.
- Let's make it possible to search by title and status on the list screen
  - [Optional requirements] If you are not a beginner, you may install a gem that makes it convenient to implement search such as ransack.
- When narrowing down on searching, let's check the changes in the issued SQL by looking into the log
  - Get in the habit of checking up logs as needed in the following steps
- Let's add on the search index on the table
- Let's add a model spec to the search (let's expand the system spec as well)

### CODE REVIEW

File: myapp/app/models/task.rb [hunk 1]
```diff
@@ -11,4 +11,17 @@ class Task < ApplicationRecord # rubocop:todo Style/Documentation
...
   validates :title, presence: true
+
+  scope :custom_order, lambda { |order_by|
+    case order_by
+    when 'latest'
+      order(created_at: :desc)
+    when 'closest-deadline'
+      order(due_date: :asc)
+    when 'far-deadline'
+      order(due_date: 'desc')
```
Comment (tana2034) [inline myapp/app/models/task.rb]: I'm curios why this is not symbol 👀 

```suggestion
      order(due_date: :desc)
```

File: myapp/app/views/tasks/index.html.erb [hunk 2]
```diff
@@ -3,8 +3,13 @@
...
 <p>
-  <%= link_to t('task.links.sort.latest'), tasks_path(sort: 'latest') %> |
-  <%= link_to t('task.links.sort.oldest'), tasks_path(sort: 'oldest') %>
+  <%= render 'search', q: @query, url: tasks_path %>
+</p>
+<p>
+  <%= link_to t('task.links.sort.latest'), tasks_path(sort: 'latest'), id: "sort_latest" %> |
+  <%= link_to t('task.links.sort.oldest'), tasks_path(sort: 'oldest'), id: "sort_oldest" %> |
+  <%= link_to t('task.links.sort.closest_deadline'), tasks_path(sort: 'closest-deadline'), id: "closest_deadline" %> |
+  <%= link_to t('task.links.sort.far_deadline'), tasks_path(sort: 'far-deadline'), id: "far_deadline" %>
```
Comment (tana2034) [inline myapp/app/views/tasks/index.html.erb:12]: oh I have seen this in previous PR.. did you merge step 12 PR to your `m-alamin` branch?

File: myapp/app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -2,10 +2,9 @@
 
 class TasksController < ApplicationController # rubocop:todo Style/Documentation
   def index
-    @tasks = Task.order(created_at: :asc)
-    return unless params[:sort] == 'latest'
+    @query = Task.ransack(params[:q])
 
-    @tasks = Task.order(created_at: :desc)
+    @tasks = @query.result(distinct: true).custom_order(params[:sort])
```
Comment (tana2034) [inline myapp/app/controllers/tasks_controller.rb]: do we need `distinct` option? 👀 is there any duplication? 🤔

**Review by tana2034 — COMMENTED** [review]
seems good 😄 I left several comments, please check them 🙇

File: myapp/app/models/task.rb (hunk 1 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/models/task.rb]: Oh its typo, sorry

File: myapp/app/views/tasks/index.html.erb (hunk 2 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/views/tasks/index.html.erb:12]: yes

File: myapp/app/controllers/tasks_controller.rb (hunk 3 above)
Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/tasks_controller.rb]: not yet though, but i think i should remove it now. i was thinking about future possibility

Comment (rr-mdalauddin-alamin) [reply myapp/app/controllers/tasks_controller.rb]: modified

Comment (tana2034) [reply myapp/app/controllers/tasks_controller.rb]: If duplication occurs, I think it is necessary to review whether the idea is correct 🤔 anyway, thank you for fixing!

File: myapp/app/views/tasks/index.html.erb (hunk 2 above)
Comment (tana2034) [reply myapp/app/views/tasks/index.html.erb:12]: did you merge the latest `m-alamin` to this branch?

Comment (rr-mdalauddin-alamin) [reply myapp/app/views/tasks/index.html.erb:12]: Yes, https://github.com/Fablic/training/pull/1593

Comment (tana2034) [reply myapp/app/views/tasks/index.html.erb:12]: ok, thanks

**Review by tana2034 — APPROVED** [review]
LGTM!
LGTM

**Pull request merged**